torch/nn/parallel/distributed.py

```diff
@@ -957,6 +960,27 @@ def no_sync(self):
         finally:
             self.require_backward_grad_sync = old_require_backward_grad_sync
 
+    @classmethod
+    def _get_active_ddp_module(cls):
+        """
+        TorchDynamo needs to know whether DDP is currently active, and access the DDP module in order to cooperatively optimize it.
+        """
+        return cls._active_ddp_module
+
+    # note, this ctxmgr function is marked 'skip' in torchdynamo, so dynamo only kicks in
+    # for the 'module_to_run' underneath
+    # see torchdynamo/eval_frame.py TorchPatcher.patch for more details
+    @contextmanager
+    def _inside_ddp_forward(self):
+        assert DistributedDataParallel._active_ddp_module is None, "Only one thread should be running DDP at a time"
```

Comment:
> Only one thread should be running DDP at a time

This restriction looks OK to me, as otherwise, collectives might desync. 

Is there any other reason that you decided to not use thread_local for `_active_ddp_module`? Will Dynamo switch threads?

Reply:
I reasoned that i did not expect anyone to use more than one thread calling into DDP and therefore just avoid complexity.  I could put the TLS thing back, it shouldn't hurt.

I am seeing this test failure though, with this assert enabled.  I think the test would still fail in the same way if I were using TLS.
https://github.com/pytorch/pytorch/runs/8007501670?check_suite_focus=true

reading the test case it's not immediately clear to me why we are entering DDP twice.  Any ideas?